docs(clouddriver): guide to configuring clouddriver with mysql (#1327)
spinnaker/spinnaker.github.io: setup/productionize/persistence/clouddriver-sql.md

---
layout: single
title:  "Set up Clouddriver to use SQL"
sidebar:
  nav: setup
---

{% include toc %}

> Clouddriver can be configured to use a MySQL compatible database in place of Redis for all of its persistence use cases. These are:

Reviewer: ```suggestion
You can configure Clouddriver to use a MySQL compatible database in place of Redis for all of its persistence use cases. These are:
```

1. Caching the state of all supported cloud resources and graphing their relationships
2. The cloud operations task repository
3. Distributed locking for the caching agent scheduler

> It is also possible to use Redis for one of the above, and SQL for another. A `DualTaskRepository` class is provided to enable online migrations between Redis and SQL in either direction without impact to running pipelines.

Reviewer: ```suggestion
You can also use Redis for one of the above, and SQL for another. A `DualTaskRepository` class is provided to enable online migrations between Redis and SQL in either direction without impact to running pipelines.
```

This guide covers database and Clouddriver configuration details, as well as tips on how Clouddriver is operated at Netflix.

## Why SQL

> Growth in Netflix's infrastructure footprint over time exposed a number of scaling challenges within Clouddriver's original Redis backed caching implementation. For example, the functional use of unordered Redis sets as secondary indexes incurs linear cost while also making it challenging to shard the keyspace without hotspots. Additionally, Clouddriver's data is highly relational in nature, making it a good candidate for modeling to either relational or graph databases.

Reviewer: ```suggestion
Growth in Netflix's infrastructure footprint over time exposed a number of scaling challenges with Clouddriver's original Redis-backed caching implementation. For example, the functional use of unordered Redis sets as secondary indexes incurs linear cost while also making it challenging to shard the keyspace without hotspots. Additionally, Clouddriver's data is highly relational, making it a good candidate for modeling to either relational or graph databases.
```

> Small Spinnaker installations are unlikely to see significant performance or cost gains by running Clouddriver on an RDBMS as opposed to Redis. In the Netflix production environment, we have seen improvements of up to 400% in the time it takes to load the clusters view for very large applications since migrating from Redis to Aurora. We also see greater consistency in 99th percentile times under peak load and have made meaningful reductions in the number of instances dedicated to operating Clouddriver. In addition to performance improvements at scale, we believe the SQL implementation is better suited to future growth and have load tested Clouddriver backed by Aurora with the Netflix Production data set at over 10x traffic rates that previously resulted in Redis related outages.

Reviewer 2: Do you want to give any indication of what "small" and "large" are?

> Small Spinnaker installations (managing up-to a few thousand instances across all applications) are unlikely to see significant performance or cost gains by running Clouddriver on an RDBMS as opposed to Redis. In the Netflix production environment, we have seen improvements of up to 400% in the time it takes to load the clusters view for large applications (hundreds of server groups with thousands of instances) since migrating from Redis to Aurora. We also see greater consistency in 99th percentile times under peak load and have made meaningful reductions in the number of instances dedicated to operating Clouddriver. In addition to performance improvements at scale, we believe the SQL implementation is better suited to future growth and have load tested Clouddriver backed by Aurora with the Netflix Production data set at over 10x traffic rates that previously resulted in Redis related outages.

Reviewer: ```suggestion
Small Spinnaker installations (managing up-to a few thousand instances across all applications) are unlikely to see significant performance or cost gains by running Clouddriver on an RDBMS as opposed to Redis. In the Netflix production environment, we have seen improvements of up to 400% in the time it takes to load the clusters view for large applications (hundreds of server groups with thousands of instances) since migrating from Redis to Aurora. We also see greater consistency in 99th percentile times under peak load and have reduced the number of instances required to operate Clouddriver. Besides improving performance at scale, we believe the SQL implementation is better suited to future growth. We've load tested Clouddriver backed by Aurora with the Netflix Production data set at more than 10x traffic rates, which previously resulted in Redis related outages.
```

At some point in the future, Spinnaker may drop support for Redis, but today it remains a good choice for evalauating Spinnaker as well as local development.

Note that cache provider classes within Clouddriver may need development to take full advantage of the `SqlCache` provider. The inital release adds secondary indexing by application specifically to accelerate `/applications/{application}/serverGroups` calls made by the UI, but only AWS caching agents and providers initially take advantage of this. Prior to adding application based indexing of AWS resources, Netflix still saw perfomrance and consistency gains using Aurora over Redis. Performance of the SQL cache provider should increase over time as Clouddriver's data access patterns evolve to better utilize features of the underlying storage.

## Configuration Considerations

> At Netflix, Orca and Clouddriver run with their own dedicated Aurora cluster but this isn't required. When sizing a database cluster for Clouddriver, it is preferable to ensure the full dataset fits within the InnoDB buffer cache. If migrating an existing Clouddriver installation from Redis to SQL, provisioning database instances with 4x the RAM of the current Redis working set should ensure a fit with room for growth. Much of the additional memory use is accounted for by secondary indexes.

Reviewer: Regarding "...it is preferable...," what happens if the full dataset doesn't fit? 
Maybe just say, "make sure" rather than "it is preferable to ensure"?

> Write latency and throughput are important considerations, especially for Spinnaker installations managing many accounts and regions. Each distinct `(cloudProvider, account, region, resourceType)` tuple generally gets its own independently scheduled caching agent instance within Clouddriver. How many of these there are for a given Clouddriver configuration can be considered the upper bound for concurrent write-heavy sessions to the database if not limited by use of the `sql.agent.maxConcurrentAgents` property. Caching agent write throughput can directly impact deployment times, especially in large environments with many accounts or multiple cloud providers. Clouddriver only attempts to write new or modified resources to the database however, moderating write requirements after initial population.

Reviewer: ```suggestion
Throughput and write latency are important considerations, especially for Spinnaker installations that manage many accounts and regions. Each distinct `(cloudProvider, account, region, resourceType)` tuple generally gets an independently scheduled caching-agent instance within Clouddriver. The number of these instances per Clouddriver configuration can be considered the upper bound for concurrent write-heavy sessions to the database, if not limited by the `sql.agent.maxConcurrentAgents` property. Caching-agent write throughput can directly impact deployment times, especially in large environments with many accounts or multiple cloud providers. Clouddriver only tries to write new or modified resources to the database however, moderating write requirements after initial population.
```

## Database Setup

Clouddriver ships with `mysql-connector-java` by default. You can provide additional JDBC connectors on the classpath such as `mariadb-connector-j` if desired. Use of a different RDBMS family will likely require some development effort at this time. Clouddriver was developed targeting Amazon Aurora's MySQL 5.7 compatible engine.

> Before deploying Clouddriver, a logical database and user grants must be manually created.

Reviewer: ```suggestion
Before you deploy Clouddriver, you need to manually create a logical database and user grants.
```

```sql
CREATE DATABASE `clouddriver` DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;

GRANT
  SELECT, INSERT, UPDATE, DELETE, CREATE, EXECUTE, SHOW VIEW
ON `clouddriver`.*
TO 'clouddriver_service'@'%'; -- IDENTIFIED BY "password" if using password based auth

GRANT
  SELECT, INSERT, UPDATE, DELETE, CREATE, DROP, REFERENCES, INDEX, ALTER, LOCK TABLES, EXECUTE, SHOW VIEW
ON `clouddriver`.*
TO 'clouddriver_migrate'@'%'; -- IDENTIFIED BY "password" if using password based auth
```

> The following MySQL configuration parameters should be considered:

Reviewer: "should be considered"...

"are recommended"? Or required?

Along those lines, if you can, remove the instances of "should" in the list of parameters.  Or make it clear what the consquences are of using something different for each one.

1. `tx_isolation` should be set to `READ-COMMITTED`
2. `tmp_table_size` should be increased if the `Created_tmp_disk_tables` MySQL metric regulary grows.
3. `sort_buffer_size` tuning may improve performance in large environments but this should be done judiciously.

> ## Configuring Clouddriver with MySQL

Reviewer: ```suggestion
## Configure Clouddriver to use MySQL
```

> The following yaml based parameters provide a Clouddriver configuration that entirely uses MySQL in place of Redis. [Halyard](/reference/halyard/) users can provide these as overrides via a `clouddriver-local.yml` file. Halyard does not yet natively support Clouddriver SQL configuration.

Reviewer: ```suggestion
The following yaml-based parameters provide a Clouddriver configuration that entirely uses MySQL in place of Redis. [Halyard](/reference/halyard/) users can provide these as overrides via a `clouddriver-local.yml` file. Halyard does not yet natively support Clouddriver SQL configuration.
```

> The following yaml based parameters provide a Clouddriver configuration that entirely uses MySQL in place of Redis. [Halyard](/reference/halyard/) users can provide these as overrides via a `clouddriver-local.yml` file. Halyard does not yet natively support Clouddriver SQL configuration.

Reviewer: Could you swap the 2nd and 3rd sentences?

    readBatchSize: 500
    writeBatchSize: 300
  scheduler:
    enabled: true
  connectionPools:
    default:
      # additional connection pool parameters are available here,
      # for more detail and to view defaults, see:
      # https://github.com/spinnaker/kork/blob/master/kork-sql/src/main/kotlin/com/netflix/spinnaker/kork/sql/config/ConnectionPoolProperties.kt
      default: true
      jdbcUrl: jdbc:mysql://your.database:3306/clouddriver
      user: clouddriver_service
      # password: depending on db auth and how spinnaker secrets are managed
    # The following tasks connection pool is optional. At Netflix, clouddriver
    # instances pointed to Aurora read replicas have a tasks pool pointed at the
    # master. Instances where the default pool is pointed to the master omit a
    # separate tasks pool.
    tasks:
      user: clouddriver_service
      jdbcUrl: jdbc:mysql://your.database:3306/clouddriver
  migration:
    user: clouddriver_migrate
  jdbcUrl: jdbc:mysql://your.database:3306/clouddriver

redis:
  enabled: false
  cache:
    enabled: false
  scheduler:
    enabled: false
  taskRepository:
    enabled: false
```

### Agent Scheduling

> The above configuration parameters result in the use a table within MySQL as part of a locking service for the caching agent scheduler. This is working well for Netflix in production against Aurora, but could result in poor overall performance for all of Clouddriver depending on the database configuration. If you observe high database CPU utilization, lock contention, or if caching agents aren't consistently running at the expected frequency, you may want to try using the Redis based scheduler to determine if this is a contributing factor. The following modification to the above example configures Clouddriver to use SQL for the cache and task repository, and Redis for agent scheduling.

Reviewer: ```suggestion
The above yaml configures Clouddriver to use a MySQL table as part of a locking service for the caching agent scheduler. This works well for Netflix in production against Aurora, but could result in poor overall performance for all of Clouddriver, depending on how your database is configured. If you observe high database CPU utilization, lock contention, or if caching agents aren't consistently running at the expected frequency, you may want to try using the Redis-based scheduler to determine if this is a contributing factor. 

The following modification to the above example configures Clouddriver to use SQL for the cache and task repository, and Redis for agent scheduling.
```

```yaml
sql:
  scheduler:
    enabled: false

redis:
  enabled: true
  connection: redis://your.redis
  cache:
    enabled: false
  scheduler:
    enabled: true
  taskRepository:
    enabled: false
```

> ### Maintaining Task Repository Availability While Migrating from Redis to SQL

Reviewer 2: Do you want to add a bit more about how to migrate? I don't think we have any docs in general on how to set up a database migration, and I think it would be a good conclusion to this guide.

Author: Good suggestion. There's no one way to do it, and I don't know how it would be done in a Halyard managed environment, so I think I'll write a "How Netflix Live Migrated Clouddriver from Redis to SQL" section.

> If migrating from Redis to SQL in a production environment where pipeline failures or downtime should be avoided, the following modifications to the first configuration example configures Clouddriver use SQL for caching and new tasks, but with fallback reads to Redis for tasks not found in the SQL database.

Reviewer: ```suggestion
If you're migrating from Redis to SQL in a production environment, where you need to avoid pipeline failures and downtime, you can modify the configuration, as shown below, to configure Clouddriver use SQL for caching and new tasks, but with fallback reads to Redis for tasks not found in the SQL database.
```

```yaml
redis:
  enabled: true
  connection: redis://your.redis
  cache:
    enabled: false
  scheduler:
    enabled: false
  taskRepository:
    enabled: true

dualTaskRepository:
  enabled: true
  primaryClass: com.netflix.spinnaker.clouddriver.sql.SqlTaskRepository
  previousClass: com.netflix.spinnaker.clouddriver.data.task.jedis.RedisTaskRepository
```

## How Netflix Migrated Clouddriver from Redis to SQL

The following steps were taken to live migrate Clouddriver in the Netflix production Spinnaker stack from Redis to SQL.

1. Provision the database. In this case, a multi-AZ Aurora cluster was provisioned with 3 reader instances.
2. Deploy Clouddriver with SQL enabled only for the `dualTaskRepository`. During the migration, traffic is split between Redis-backed Clouddriver instance and SQL-backed instances. It is important that tasks can be read regardless of request routing to avoid pipeline failures.

>     ```yaml

Reviewer: @asher 
Might be helpful to identify which config file this is.

      enabled: true
      primaryClass: com.netflix.spinnaker.clouddriver.sql.SqlTaskRepository
      previousClass: com.netflix.spinnaker.clouddriver.data.task.jedis.RedisTaskRepository
    ```
4. A clouddriver-sql-migration-caching server group was deployed with the above configuration, followed by a 5 minute `WaitStage`, allowing time for cache population. Clouddriver API server groups (with caching agent execution disabled) were then deployed behind the same load balancers routing traffic to the Redis-backed server groups.
5. After another 5 minute wait, the Redis-backed server groups were disabled. At this time, any tasks running on the disabled Redis instance continue until finished. All new requests are routed to clouddriver-sql-migration-api server groups. If a requested taskId is not present in the SQL database, Clouddriver attempts to read it from Redis.
6. The clouddriver-sql-migration configuration is merged into main, with the following changes:
    ```yaml
    redis:
      enabled: false

    dualTaskRepository:
      enabled: false
    ```
7. The disabled Redis-backed Clouddriver instances were verified as idle and the new configuration deployed via a red/black.

> 8. The temporary migration clusters were disabled and then terminated 5 minutes later.

Reviewer 2: This section is great! Thank you @asher
